docs/adr/0003-threadsafe-interfaces.md

```diff
@@ -0,0 +1,196 @@
+would grow a corresponding boolean `threadsafe` field and corresponding public accessor.
+When [building an instance of this struct from the UDL](https://github.com/mozilla/uniffi-rs/blob/803bb3d79daa8ea088fb2d8f05c08ada09821986/uniffi_bindgen/src/interface/mod.rs#L786),
+we would inspect the list of attributes to look for one named "Threadsafe" and set the field
+if present. The way that we [handle the `[ByRef]` annotation on method arugments](https://github.com/mozilla/uniffi-rs/blob/803bb3d79daa8ea088fb2d8f05c08ada09821986/uniffi_bindgen/src/interface/mod.rs#L658)
+will likely serve as a good example to follow.
+
+### Implement a less-locking HandleMap variant
+
+The `ffi_support` crate provides a basic non-locking [`HandleMap<T>`](https://docs.rs/ffi-support/0.4.0/ffi_support/handle_map/struct.HandleMap.html) struct,
+and it implements `ConcurrentHandleMap` as a thin wrapper around a `RwLock<HandleMap<Mutex<T>>>`. We can make our own variant that removes the inner mutex, as a thin wrapper around a`RwLock<HandleMap<T>>`. (The outer `RwLock` is still needed in order to guard mutations of the handlemap itself, and we must be careful to only hold it for short periods).
+
+Bikeshed name: `NonLockingHandleMap`, implemented under the [`uniffi::ffi` module](https://github.com/mozilla/uniffi-rs/tree/main/uniffi/src/ffi). It needs to live in the `uniffi` crate so that it can be used by the generated Rust scaffolding at runtime.
+
+We don't need to support the entire `ConcurrentHandleMap` interface, only:
+* `insert_with_output`
+* `insert_with_result`
+* `call_with_output`
+* `call_with_result`
+* `delete_u64`
+
+The `NonLockingHandleMap<T>` struct should be capable of handing out `&T` references but should never hand
+out a `&mut T`. It should also require that `T: Send + Sync`.
+
+### Use `NonLockingHandleMap` in the Rust scaffolding for `[Threadsafe]` instances
+
+In [ObjectTemplate.rs](https://github.com/mozilla/uniffi-rs/blob/main/uniffi_bindgen/src/templates/ObjectTemplate.rs), check the `threadsafe` property of the object definition. If it's true, use `NonLockingHandleMap` instead of `ConcurrentHandleMap` in the [lazy static that declares the handlemap for that interface](https://github.com/mozilla/uniffi-rs/blob/803bb3d79daa8ea088fb2d8f05c08ada09821986/uniffi_bindgen/src/templates/ObjectTemplate.rs#L11).
+
+By ensuring that `NonLockingHandleMap` and `ConcurrentHandleMap` expose a similar API, this will hopefully
+be a fairly minimal if-then-else that just chooses the correct name of the struct to use.
+
+### Optional: Add a `[Blocking]` annotation to methods and functions
+
+As a nod toward a possible future where uniffi can provide more help in dealing with blocking vs
+async I/O, we could also support annotating individual methods/functions to indicate whether they
+may block. We don't have concrete plans to *use* this information right now, but it could serve
+as useful documentation to consumers of the interface.
```

Comment:
I would assert that we don't really want this to be optional, in the sense that I think there's high value in (at the very least) encouraging method implementors to document if they're going to make something blocking as part of their contract with the API consumer.

That said, I almost want to suggest different semantics: have a developer affordance to require the developer to explicitly think through and document what their intent is at the time the code is written.  Which is to say, have [Blocking] and [NonBlocking] attributes, and disallow attribute-free methods.

Reply:
Oh, interesting suggestion. I think I'd be OK with trying that out as a required annotation and see if we like it, as we can always remove the restriction later.

Reply:
Indeed; that's the nice thing about only having one real client -- it's still pretty easy to experiment with stuff like this.  :-)

Reply:
I guess constructors would need to be declared as [Blocking], since there is gonna be some locking around the handlemap...

Reply:
I'm not sure what we intend the semantics of `[Blocking]` to mean, but subjectively, there feels like a significant difference between "blocks on network or disk I/O" and "takes a mutex protecting a purely-in-memory data structure". *Any* method call on an object is going to have to interact with the handlemap mutex, so I worry that considering handlemap access a blocking operation will eliminate any documentary value that the annotations may provide.

Reply:
By contrast, I think there *is* value in being able to know whether or not the constructor does disk I/O via some sort of annotation.

Reply:
I agree with all the above, but in the short term, I believe the main value is documenting expectations, so would like to see *something* here sooner rather than later (eg, we will have `getExperimentBranch()` which doesn't "block", but now there's a `getExperimentBranches()` function which, IIUC, does *not* need these semantics - which will not be at all obvious without some url annotation